daemon/logger/loggerutils/follow.go

```diff
@@ -1,211 +1,162 @@
 			if err != nil {
-				return err
+				fl.Watcher.Err <- fmt.Errorf("logger: error opening log file for follow after rotation: %w", err)
+				return
+			}
+
+			if nrot := wrote.rotation - read.rotation; nrot > 1 {
+				fl.log.WithField("missed-rotations", nrot).
+					Warn("file rotations were missed while following logs; some log messages have been skipped over")
 			}
-			fl.retries++
-			return errRetry
+
+			// Set up our read position to start from the top of the file.
+			read.size = 0
+		}
+
+		if fl.decode(io.NewSectionReader(f, read.size, wrote.size-read.size)) {
+			return
 		}
-		return err
-	case <-fl.logWatcher.WatchProducerGone():
-		return errDone
-	case <-fl.logWatcher.WatchConsumerGone():
-		return errDone
+		read = wrote
 	}
 }
 
-func (fl *follow) handleDecodeErr(err error) error {
-	if !errors.Is(err, io.EOF) {
-		return err
+// nextPos waits until the write position of the LogFile being followed has
+// advanced from current and returns the new position.
+func (fl *follow) nextPos(current logPos) (next logPos, ok bool) {
+	var st logReadState
+	select {
+	case <-fl.Watcher.WatchConsumerGone():
+		return current, false
+	case st = <-fl.LogFile.read:
 	}
 
-	// Handle special case (#39235): max-file=1 and file was truncated
-	st, stErr := fl.file.Stat()
-	if stErr == nil {
-		size := st.Size()
-		defer func() { fl.oldSize = size }()
-		if size < fl.oldSize { // truncated
-			fl.file.Seek(0, 0)
-			fl.dec.Reset(fl.file)
-			return nil
-		}
-	} else {
-		logrus.WithError(stErr).Warn("logger: stat error")
+	// Have any any logs been written since we last checked?
+	if st.pos == current { // Nope.
+		// Add ourself to the notify list.
+		st.wait = append(st.wait, fl.c)
+	} else { // Yes.
+		// "Notify" ourself immediately.
+		fl.c <- st.pos
 	}
+	fl.LogFile.read <- st
 
-	for {
-		err := fl.waitRead()
-		if err == nil {
-			break
-		}
-		if err == errRetry {
-			continue
+	select {
+	case <-fl.LogFile.closed: // No more logs will be written.
+		select { // Have we followed to the end?
+		case next = <-fl.c: // No: received a new position.
+		default: // Yes.
+			return current, false
 		}
-		return err
+	case <-fl.Watcher.WatchConsumerGone():
+		return current, false
+	case next = <-fl.c:
 	}
-	return nil
+	return next, true
 }
 
-func (fl *follow) mainLoop(since, until time.Time) {
+// decode decodes log messages from r until EOF and sends them to the log
+// watcher.
+func (fl *follow) decode(r io.Reader) (done bool) {
+	fl.Decoder.Reset(r)
 	for {
-		select {
-		case err := <-fl.notifyEvict:
-			if err != nil {
-				fl.handleMustClose(err.(error))
-			}
-			return
-		default:
-		}
-		msg, err := fl.dec.Decode()
+		msg, err := fl.Decoder.Decode()
 		if err != nil {
-			if err := fl.handleDecodeErr(err); err != nil {
-				if err == errDone {
-					return
-				}
-				// we got an unrecoverable error, so return
-				fl.logWatcher.Err <- err
-				return
+			if errors.Is(err, io.EOF) {
+				return false
```

Comment:
maybe I misunderstood, but why return `done=false` on EOF?

Reply:
`done=true` signals that a terminal stopping condition for following logs has been encountered while decoding: unexpected error, watch consumer is gone, or the `docker logs --until` condition has been reached. EOF means that the follower has caught up to the log writer or has reached the end of a rotated log file, neither of which is a reason to stop following.